Main.hs

```diff
@@ -1,39 +1,106 @@
+{-# LANGUAGE InstanceSigs #-}
+
 module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Num Double
+  | Sqrt Expr
+  | Minus Expr Expr
+  | Plus Expr Expr
+  | Multi Expr Expr
+  | Div Expr Expr
+  | Power Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = printf "sqrt %s" (show a)
+  show (Minus a b) = printf "%s - %s" (show a) (show b)
+  show (Plus a b) = printf "%s + %s" (show a) (show b)
+  show (Multi a b) = printf "%s" (show a) (show b)
+  show (Div a b) = printf "%s / %s" (show a) (show b)
+  show (Power a b) = printf "%s ^ %s" (show a) (show b)
 
 instance Eq Expr where
-  (==) = undefined
+  (==) :: Expr -> Expr -> Bool
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  Minus a b == Minus c d = a == c && b == d
+  Plus a b == Plus c d = a == c && b == d
+  Multi a b == Multi c d = a == c && b == d
+  Div a b == Div c d = a == c && b == d
+  Power a b == Power c d = a == c && b == d
+  _ == _ = False
 
 data Error
+  = DivisionByZero Expr Expr
+  | SqrtOfNegNumber Expr
 
 instance Show Error where
-  show = undefined
+  show (DivisionByZero a b) = printf "Error: tried to do devision by zero: %s / %s" (show a) (show b)
+  show (SqrtOfNegNumber a) = printf "Error: tried to find the square root of a negative number: sqrt %s" (show a)
 
 instance Eq Error where
-  (==) = undefined
+  DivisionByZero _ _ == DivisionByZero _ _ = True
+  SqrtOfNegNumber _ == SqrtOfNegNumber _ = True
+  _ == _ = False
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Num a) = Right a
+eval (Sqrt a) = do
+  val <- eval a
+  if val >= 0 then Right (sqrt val) else Left (SqrtOfNegNumber a)
+eval (Minus a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  Right (val1 - val2)
+eval (Plus a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  Right (val1 + val2)
+eval (Multi a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  Right (val1 * val2)
+eval (Div a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  if val2 == 0
+    then Left (DivisionByZero a b)
+    else Right (val1 / val2)
+eval (Power a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  Right (val1 ** val2)
```

Comment:
You have a lot of code duplication here. Could you introduce a helper function to get rid of it?

Reply:
added the helper function and changed the code according the previous comments